### PR TITLE
Restructuring of deepquest for packaging

### PR DESCRIPTION
**Pull request by davidwyld**
This PR incorporates a major restructuring of the project to:
1. Make it a pip installable package.
2. Remove unnecessary components of the codebase, mainly NMT-Keras, to use the master repo of that project (lvapeab/nmt-keras)
    - This included making PRs to lvapeab/nmt-keras in order to package that project to be usable by deepquest.
3. Move `dq_tests` to `tests` and `dq_utils` to `deepquest/utils`.
4. Move the models directory to `deepquest/qe_models`
5. Collate all data preparation functions into `deepquest/utils/prepare_data.py` including `preprocessDoc`.

Closes https://github.com/RSE-Sheffield/deepquest/issues/28 and closes https://github.com/RSE-Sheffield/deepquest/issues/3

At present, in order to run deepquest, it must be installed as a package either by:
1. Cloning the repo and running `pip install .` for `setup.py`, or
2. Installing from the repo directly as `pip install 'deepquest @ https://github.com/rse-sheffield/nmt-keras/archive/master.zip'`
    - Note that this option will not copy all files from the repo, but only those required for running the package.

deepquest can be run from any directory, after installation, either from the command line as one of the following:
```shell
deepquest train -c configs/config.yml
dq train -c configs/config.yml
```
or via importing as a python module, for instance:
```python
import deepquest as dq
dq.train('configs/config.yml')
```

Currently, deepquest will look in the present working directory for all data and config files, and put the `trained_models` directory there also.

**Note**: as a python module, seed setting is not implemented, as this must happen before importing any of the modules that use the seed. So, for deterministic results, one would need to do the following before importing deepquest.
```python
seed = 1
import numpy.random
numpy.random.seed(seed)
import random
random.seed(seed)
```

### CODE REVIEW

File: deepquest/predict.py [hunk 1]
```diff
@@ -247,7 +247,7 @@ def main(model, dataset, save_path=None, evalset=None, changes={}):
 
     # from nmt_keras import model_zoo
     from keras.utils import CustomObjectScope
-    import nmt_keras.models.utils as layers  # includes all layers and everything defined in nmt_keras.utils
+    import deepquest.qe_models.utils as layers  # includes all layers and everything defined in nmt_keras.utils
```
Comment (drj11) [inline deepquest/predict.py]: comment needs updating to reflect where layers come from

File: deepquest/train.py [hunk 2]
```diff
@@ -138,7 +138,7 @@ def train_model(params, weights_dict, load_dataset=None, trainable_pred=True, tr
             # otherwise we just reload the weights
             # from the files containing the model
             from keras.utils import CustomObjectScope
-            import nmt_keras.models.utils as layers  # includes all layers and everything defined in nmt_keras.utils
+            from deepquest.qe_models import utils as layers  # includes all layers and everything defined in nmt_keras.utils
```
Comment (drj11) [inline deepquest/train.py]: same as for predict, comment needs updating.

File: deepquest/train.py [hunk 3]
```diff
@@ -320,21 +320,21 @@ def main(config=None, changes={}):
     :param changes: Optional dictionary of parameters to overwrite config.
     """
     if isinstance(config, str):
-        if arg.endswith('.yml'):
+        if config.endswith('.yml'):
             # FIXME make this a user option (maybe depend on model type and level?)
             with open('configs/default-config-BiRNN.yml') as file:
```
Comment (drj11) [inline deepquest/train.py:325]: I'll come over and ask you about this.

File: deepquest/utils/prepare_data.py [hunk 4]
```diff
@@ -9,6 +9,74 @@
 # We patch the Dataset class here.
 # This will not be necessary when multimodal_keras_wrapper is fixed.
```
Comment (drj11) [inline deepquest/utils/prepare_data.py]: This comment is supposed to be attached to, and precede, the `dataset_preprocessreal_fixed` function. But you slapped a gigantic new function in between this comment and the function that follows. I suppose that's my fault for putting blank lines between them.

File: setup.py [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+    long_description=long_description,
+    long_description_content_type="text/markdown",
+    url="https://github.com/RSE-Sheffield/deepquest/",
+    classifiers=[
+        "Programming Language :: Python :: 3",
+        "Operating System :: OS Independent",
+    ],
+    install_requires=[
+        "future",
+        "numpy<1.17,>=1.14.5",
```
Comment (drj11) [inline setup.py]: that's slightly surprising and might need a comment to justify it

**Review by drj11 — COMMENTED** [review]
cool!

File: setup.py (hunk 5 above)
Comment (davidwyld) [reply setup.py]: Note to self: is this not set in the TensorFlow requirements?

File: deepquest/utils/prepare_data.py (hunk 4 above)
Comment (davidwyld) [reply deepquest/utils/prepare_data.py]: Quite right! Have actually just removed `dataset_preprocessreal_fixed` in a6c9376 since this is now fixed in multimodal_keras_wrapper.

File: setup.py (hunk 5 above)
Comment (davidwyld) [reply setup.py]: There was an incompatibility warning between tensorflow and numpy, but this has now been resolved as discussed here https://github.com/tensorflow/tensorflow/pull/30559#issuecomment-518258441
Have removed these requirements here.
